design/nfs/nfs-controller-runtime.md

```diff
@@ -0,0 +1,161 @@
+- **Envtest:** Provides libraries for integration testing by starting a local control plane (etcd and kube-apiserver).
+- **Matrics:** Provides metrics utility for controller.
+
+## Implementation
+
+The implementation of this proposal is to rewrite NFS Operator controller to use controller-runtime and introduce the validation admission webhook using controller-runtime for NFS Operator.
+
+### Controller & Reconciliation
+
+Operators are Kubernetes extensions that use custom resources to manage applications and their components using the Kubernetes APIs and kubectl tooling. Operators follow the Kubernetes controller principles. The process in which the actual state of the object (both cluster object and external object) will be matching the desired state which called *Reconciliation* process in the controller-runtime. 
+
+The current implementation is the operator watch an event (create, update and delete) of CustomResource and will be handled by registered function in `ResourceEventHandlerFuncs` which every event has its own handler but only the create handler that implemented.
+
+Controller-runtime introduces an interface called `Reconciler` that will ensure the state of the system matches what is specified by the user in the object at the time the `Reconciler` is called. `Reconciler` responds to generic events so it will contain all of the business logic of a Controller (create, update, and delete). The controller-runtime also have utility functions for creating and updating an object called `CreateOrUpdate` which will make easier to handling update of an object.
+
+Since this implementation only changes the logic of the controller, so the deployment process will be like current implementation.
+
+### Validation
+
+CustomResource validation in the operator can be done through the Controller itself. However, the operator pattern has two common types to validate the CustomResource.
+
+- **Syntactic validation** By defining OpenAPI validation rules.
+- **Semantic Validation** By creating a ValidatingAdmissionConfiguration and Admission Webhook.
+
+The current implementation only validates the CustomResource in the controller and just gives an error log in the operator stdout if the given resource is invalid. In this implementation will also cover the CustomResouce validation both though `Syntactic validation` and `Semantic Validation` and also give an improvement validation in the controller.
+
+To implement `Syntactic validation` is only by defining OpenAPI validation rules. Otherwise, the `Semantic Validation` implementation is a bit more complicated. Fortunately, controller-runtime provides an awesome package to help to implement Admission Webhook. Since the webhook server must be served through TLS, a valid TLS certificate will be required. In this case, we can depend on [cert-manager](https://cert-manager.io/)
```

Comment:
one of the things i'm particularly interested in is how this work to get webhooks working for NFS can also be reused and leveraged for other storage providers in Rook.  Is there general (reusable) functionality, helpers, documentation, etc. that we can provide as part of this effort for NFS that will also be helpful for all the the other storage providers to simply reuse and not have to solve the same problem again that we are solving here for NFS?

This makes Rook all the more strong as a *framework* for multiple providers :)